src/docs/arch-block.dox

```diff
@@ -1,9 +1,192 @@
+Any blocks allocated by the block manager must be a multiple of this value.
+
+For example, if we specify an allocation_size of \c 4KB, blocks of size
+\c 8KB and \c 12KB would be permitted but NOT \c 10KB. Higher allocation_sizes
+may yield better compression rates, yet may cause larger page sizes. The
+allocation_size is set to \c 4KB by default which is a good choice unless your
+OS or storage device has special requirements.
+
+@subsection checksum Checksum
+
+The \c checksum configuration controls whether blocks have a checksum stored
+alongside them. With block checksums enabled, a block written to disk it is
+accompanied by a checksum of the block contents. When that block is read off the
+disk to reconstruct a page, we recompute the checksum and verify that is the
+same to defend against the risk of file corruption.
+
+@section block_usage How WiredTiger uses the block manager
+
+@subsection creation File creation and the block manager
+
+When a new file is created in WiredTiger via WT_SESSION::create, the file is
+created on disk and the associated \c allocation_size is written out to the
+metadata file. However the block manager itself isn't created. Only when a btree
+is opened does WiredTiger allocate a block manager. Each btree has it's own
+block manager.
+
+@subsection read Reading files and pages
+
+When an existing btree is opened for the first time, the location of the root
+block is contained in the metadata file \c WiredTiger.wt. The block manager will
+read the block at the location specified and return the page image as a buffer
+to the layer above. This will then be instantiated as a page in memory.
+
+From there subsequent page addresses can be read from the root page and the
+process repeated as required. If a cursor traverses to a page which hasn't been
+read into memory the same process will take place.
+
+@subsection Writing
+
+Two cases exist for writing out data using the block manager, checkpoint and
+eviction. Both are explained in the following two sections. When a page image
+is written out the block manager the \c bm->write API is called. See \c bt_io.c
+for more detail.
+
+@subsubsection Checkpoint
+
+Checkpoints are the process by which a file can be guaranteed to contain a
+consistent snapshot of the data at a given point in time for more detail on that
+see @ref arch-checkpoint.
+
+At the block manager level a checkpoint corresponds with a set of blocks which
+are stored in the file associated with the given URI. Typically each file will
+contain a minimum of two checkpoints. When opening an existing file the latest
+checkpoint is read.
+
+When WiredTiger performs a checkpoint clean pages will not result in additional
+blocks being created and as such existing blocks from older checkpoints may be
+reused, this means that while there may be more than one checkpoint in a file
+the block are not guaranteed to be mutually exclusive. Checkpoints are created
```

Comment:
To me "mutually exclusive" is just set theory, i.e. the set of the blocks in each checkpoint aren't exclusive. I can re-word it a bit.

Reply:
I don't think "mutually exclusive" suggests concurrency, but I do think it makes the explanation unnecessarily complicated. Can we say something like this:

_When performing a checkpoint, new blocks are only written for dirty pages. The same block can be included in more than one checkpoint, such as when a page has not been modified since the last checkpoint occurred._

An observation - there are a LOT of sentences starting with "When ..."

Reply:
Suggestions on how to not start sentences with When?

Reply:
Switch the sentence around, e.g.

_When performing a checkpoint, new blocks are only written for dirty pages._  --> New blocks are only written for dirty pages when performing a checkpoint._

_When opening an existing file the latest checkpoint is read._ --> _We read the latest checkpoint when opening an existing file._

Use a different word, e.g.

_When performing a checkpoint, new blocks are only written for dirty pages._  --> _During a checkpoint, new blocks are only written for dirty pages._ 

_When opening an existing file the latest checkpoint is read._ --> _Upon opening an existing file the latest checkpoint is read._ 

Don't use it if you don't need it, e.g.

_When operating in first fit mode the block manager will place the newly created block in the first available extent._ --> _In first fit mode the block manager places the newly created block in the first available extent._

_When writing the root node of a b-tree first fit mode is used._ --> _We use first fit mode to write the root node of a btree._

Reply:
I tried to explaining it explicitly rather than vaguely in commit 8a13dec. I also de-when'd the suggested locations.

Reply:
Nice! I really like how you first stated the point you wanted to make and then followed up with an example in the rewording.